src/integrationtests/java/com/aws/greengrass/integrationtests/logmanager/LogManagerTest.java

```diff
@@ -453,4 +461,174 @@ void GIVEN_system_config_with_small_periodic_interval_WHEN_active_logs_included_
         assertEquals(1, allFiles.size());
         logManagerService.ACTIVE_LOG_FILE_FEATURE_ENABLED_FLAG.set(false);
     }
+
+    @Test
+    void GIVEN_files_randomly_named_WHEN_all_files_renamed_randomly_THEN_all_files_only_send_once()
+            throws Exception {
+        //TODO: Intentionally pass the test now, since it will fail for current code.
+        if (!ACTIVE_LOG_FILE_FEATURE_ENABLED_FLAG.get()) {
+            return;
+        }
+        when(cloudWatchLogsClient.putLogEvents(any(PutLogEventsRequest.class))).thenReturn(
+                PutLogEventsResponse.builder().nextSequenceToken("nextToken").build());
+
+        tempDirectoryPath = Files.createDirectory(tempRootDir.resolve("logs"));
+        LogConfig.getRootLogConfig().setLevel(Level.TRACE);
+        LogConfig.getRootLogConfig().setStore(LogStore.FILE);
+        LogManager.getLogConfigurations().putIfAbsent("UserComponentB",
+                newLogConfigFromRootConfig(LogConfigUpdate.builder().fileName("UserComponentB.log").build()));
+
+        List<LogFile> logFiles = new ArrayList<>();
+        int testFileNumber = 4;
+        for (int i = 0; i < testFileNumber; i++) {
+            String randomName1 = UUID.randomUUID().toString();
+            String randomName2 = UUID.randomUUID().toString();
+            LogFile logFile = createTempFileAndWriteDataAndReturnFile(tempDirectoryPath,
+                    randomName1 + "UserComponentB.", randomName2);
+            logFiles.add(logFile);
+        }
+
+        setupKernel(tempDirectoryPath, "randomlyNamedFilesWithoutComponentsConfig.yaml");
+        //TODO: this lazy sleeping inherits from existing tests, but should be replaced by some better mechanism in
+        // future refactoring
+        TimeUnit.SECONDS.sleep(15);
+
+        // randomly rename file and see if uploader trying to process more file.
+        for (LogFile logFile : logFiles) {
+            String randomName3 = UUID.randomUUID().toString();
+            String randomName4 = UUID.randomUUID().toString();
+            LogFile renameFile = new LogFile(Files.createTempFile(tempDirectoryPath,
+                    randomName3 + "UserComponentB.", randomName4).toUri());
+            String tempFileHash = logFile.hashString();
+            logFile.renameTo(renameFile);
+            assertEquals(renameFile.hashString(), tempFileHash);
+            assertFalse(logFile.exists());
+        }
+        //TODO: this lazy sleeping inherits from existing tests, but should be replaced by some better mechanism in
+        // future refactoring
+        TimeUnit.SECONDS.sleep(15);
+
+        // Even the file are all renamed, the cloudWatchClient should only try to upload events once, and the total
+        // size should be contents of 4 files.
+        verify(cloudWatchLogsClient, times(1)).putLogEvents(captor.capture());
```

Comment:
Also verify that our map of filehashes has a size of 4. I want to make sure that we aren't growing that map forever.

Reply:
We do not keep the global map of fileHash anymore. The new structure is introduced to scan and filter all files in directory, which is a kinda snapshot mechanism of the current directory.

Reply:
There is some map somewhere though, right? otherwise you'll lose the information about what we've uploaded and then we'll be duplicating the logs in the upload.

So, where are we tracking the hash now?

Reply:
Avoiding duplicate logs is done by the lastModifiedTime and fileHash of the file that lastly uploaded, which are stored in runtime config and can be seen in next PR. 
For every scanning, we will create a logFileGroup structure and filter out uploaded and rotated file, so the internal map of fileHash is newly created. 
Here is the code: https://github.com/aws-greengrass/aws-greengrass-log-manager/blob/main/src/main/java/com/aws/greengrass/logmanager/model/LogFileGroup.java#L56

The logic is: 
1. Get the timestamp of the moment of last upload, and the fileHash of the partially reading file. 
2. Create the logFIleGroup, which will scan all files in the directory, filter files that is before the timestamp of last upload. Then store the Map <fileHash, file> as internal map. 
3. get the startPosition based on matched fileHash. 
4. After uploading, re-create the logFileGroup to get the fresh snap-shot of the directory, then classify file status based on fileHash.